service/src/main/java/bio/terra/externalcreds/dataAccess/GA4GHVisaDAO.java

```diff
@@ -48,7 +49,7 @@ public GA4GHVisa insertVisa(GA4GHVisa visa) {
     var generatedKeyHolder = new GeneratedKeyHolder();
     jdbcTemplate.update(query, namedParameters, generatedKeyHolder);
 
-    return visa.withId(generatedKeyHolder.getKey().intValue());
+    return ImmutableGA4GHVisa.copyOf(visa).withId(generatedKeyHolder.getKey().intValue());
```

Comment:
My understanding is that the `withId`method by itself already creates a copy (screenshot below). Is the `ImmutableGA4GHVisa.copyOf(...)` method used here because just using `withId` would output the wrong (mutable) object type? Or is there some other reason for using this? I want to make sure I understand so I know when and where to use `copyOf` in the future. 

<img width="485" alt="image" src="https://user-images.githubusercontent.com/23193756/127875537-2ac1dda3-df55-4988-83d4-d24aa36a5886.png">

Reply:
the type of `visa` here is `GA4GHVisa`, not `ImmutableGA4GHVisa`. The `GA4GHVisa` interface does not have a `withId` function so we need to copy it to get an instance of `ImmutableGA4GHVisa` (which if you look at the implementation of `copyOf` does not actually copy if it doesn't have to)

Reply:
Okay great, that makes sense! Thanks for the explanation.